### PR TITLE
feat(auth): Save and migrate sign in method

### PR DESCRIPTION
**Pull request by tylerjroach**
- [ ] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*

*How did you test these changes?*
(Please add a line here how the changes were tested)

- [ ] Added Unit Tests
- [ ] Added Integration Tests

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by gpanshu — APPROVED** [review]

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/testutils/CredentialStoreUtil.kt [hunk 1]
```diff
@@ -28,12 +28,18 @@ import java.util.Date
...
+            "1234567890",
+            "amplify_user",
             Date(0),
-            SignInMethod.SRP,
+            SignInMethod.ApiBased(SignInMethod.ApiBased.AuthType.USER_SRP_AUTH),
             DeviceMetadata.Metadata("someDeviceKey", "someDeviceGroupKey", "someSecret"),
-            CognitoUserPoolTokens("idToken", "accessToken", "refreshToken", 1212),
+            CognitoUserPoolTokens(
+                "idToken",
+                "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwidXNlcm5hbWUiOiJhbXBs" +
```
Comment (div5yesh) [inline aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/testutils/CredentialStoreUtil.kt:38]: can we move this to top of class as a const?

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AmplifyCredential.kt [hunk 2]
```diff
@@ -21,17 +21,28 @@ import kotlinx.serialization.Serializable
 
 @Serializable
 sealed class AmplifyCredential {
+
+    interface UserPoolData {
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AmplifyCredential.kt:25]: nit: should we name this as UserPoolTypeCredential and IdentityPoolTypeCredential?

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/SignInMethod.kt [hunk 3]
```diff
@@ -15,6 +15,25 @@
...
-    SRP, CUSTOM, HOSTED
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+sealed class SignInMethod {
+
+    @Serializable
+    @SerialName("SignInMethod.ApiBased")
+    data class ApiBased(val authType: AuthType) : SignInMethod() {
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/SignInMethod.kt:26]: 🙌

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/data/AWSCognitoLegacyCredentialStore.kt [hunk 4]
```diff
@@ -149,12 +158,32 @@ class AWSCognitoLegacyCredentialStore(
...
+        val tokenUsername =
+            try {
+                cognitoUserPoolTokens.accessToken?.let { SessionHelper.getUsername(it) } ?: ""
+            } catch (e: Exception) {
+                ""
+            }
 
-        return if (cognitoUserPoolTokens == null) null
-        else SignedInData("", username ?: "", Date(0), SignInMethod.SRP, deviceMetaData, cognitoUserPoolTokens)
+        return SignedInData(
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/data/AWSCognitoLegacyCredentialStore.kt:179]: TODO: Follow up coming soon in federation to identity pool API where signed in data will be null.

**Review by div5yesh — APPROVED** [review]
LGTM

**Comment by tylerjroach** [comment]
@div5yesh I had squash merge on, but I will make all of the suggested updates in the follow up federated PR

**Pull request merged**